File: leetcode/Dynamic Programming/house_robber_II.c
// https://leetcode.com/problems/house-robber-ii/

#define MAX(a,b) (a) > (b) ? (a) : (b)

int rob1(int* nums, int numsSize) {
    int res[numsSize + 2];

    memcpy(res, nums, sizeof(int) * numsSize);
    res[numsSize] = 0;
    res[numsSize+1] = 0;

    for (int i = 2; i < numsSize + 2; i++) {
        if (i == 2) {
            res[i] += res[i-2];
            continue;
        }
        res[i] += MAX(res[i-2], res[i-3]);
    }

    return res[numsSize+1];
}


int rob(int* nums, int numsSize) {
    if (numsSize == 1) {
        return nums[0];
    }

    int res = rob1(nums + 1, numsSize - 1);
    nums[numsSize-1] = 0;
    res = MAX(res, rob1(nums, numsSize));

    return res;
}

File: leetcode/Dynamic Programming/house_robber_II.erl
%% https://leetcode.com/problems/house-robber-ii/

-module(house_robber_II).

-compile(export_all).

-spec rob(Nums :: [integer()]) -> integer().
rob(_Nums = [H | _]) when length(_Nums) == 1 ->
    H;
rob(_Nums = [H | T]) ->
    max(
        rob(T, []),
        rob(drop_last_element(T), [H])
    ).

drop_last_element(A) -> lists:reverse(tl(lists:reverse(A))).

rob([], [V1 | [V2 | _]]) ->
    max(V1, V2);
rob([], [V1 | _]) ->
    V1;
rob(_Nums = [H | T], Arr) ->
    NextVal = case Arr of
        [] -> H; %% empty arr
        [_ | [V1 | [V2 | _]]] -> H + max(V1, V2);  %% >3 elements in arr
        [_ | [V1 | _]] -> H + V1;  %% 2 elements in arr
        [_ | _] -> H  %% 1 element in arr
    end,
    rob(T, [NextVal | Arr]).
